games/achievements.py

```diff
@@ -42,101 +55,152 @@ class DNFAchievement(Achievement):
     description = "Participated in a game that completed, where you didn't"
     icon = "coffin.svg"
 
-    def has_achieved(user):
-        return (
+    def get_level(user):
+        if (
             user.gameplayer_set.filter(dnf=True)
             .filter(game__dnf=False, game__end_datetime__isnull=False)
             .exists()
-        )
+        ):
+            return AchievementLevel.GOLD
+        return AchievementLevel.NO_LEVEL
 
 
-class Top10Achievement(Achievement):
-    name = "Top 10"
-    description = "Placed top 10 total sips in a season"
+class TopAchievement(Achievement):
+    name = "Top"
+    description = "Placed top (10/5/3/1) total sips in a season"
     icon = "trophy-cup.svg"
 
-    def has_achieved(user):
+    def get_level(user):
         current_season = Season.current_season()
+        highest_rank = AchievementLevel.NO_LEVEL
+
         for i in range(1, current_season.number):  # Exclude current season
             top10 = (
                 PlayerStat.objects.filter(season_number=i)
                 .order_by("-total_sips")[:10]
                 .values("user")
             )
-            if {"user": user.id} in top10:
-                return True
-
-        return False
+            top5 = top10[:5]
+            top3 = top10[:3]
+            top = top10[:1]
```

Comment:
```suggestion
            try:
                rank = top10.index({"user": user.id})
            except ValueError:
                continue
```

and then use max on `highest_rank`.

Reply:
I generally try to avoid exceptions for control flow, so I would propose changing the function to:
```
    def get_level(user):
        current_season = Season.current_season()
        highest_rank = 11 # Should always resolve to NO_LEVEL

        for i in range(1, current_season.number):
            query = (PlayerStat.objects.filter(season_number=i)
                .order_by("-total_sips")
                .values("user")
            )

            for idx, value in enumerate(query[:10]):
                if value == {"user": user}:
                    highest_rank = min(idx + 1, highest_rank)
                    break

        if highest_rank == 1:
            return AchievementLevel.GOLD
        elif highest_rank <= 3:
            return AchievementLevel.SILVER
        elif highest_rank <= 5:
            return AchievementLevel.BRONZE
        elif highest_rank <= 10:
            return AchievementLevel.BASE
        else:
            return AchievementLevel.NO_LEVEL
```

Reply:
I think using exception handling for builtins in python is fine, as that is just how the language is designed...